tests/Unit/ChangelogTest.php

```diff
@@ -0,0 +1,268 @@
+
+                $needEmptyLine = true;
+
+                continue;
+            }
+
+            if (0 === strpos($line, '### ')) {
+                self::assertTrue($wasEmptyLine, 'CHANGELOG line ' . ($index - 1) . ' MUST be empty');
+                self::assertTrue($inTitle2, 'CHANGELOG line ' . $index . ' MUST have title2 inside a title2');
+                self::assertFalse($title3Empty, 'CHANGELOG line ' . $index . ' MUST not have empty title3');
+                self::assertMatchesRegularExpression('/^### (Fixed|Added|BC-BREAK)$/', $line, 'CHANGELOG line ' . $index . ' MUST be "Fixed", "Added" OR "BC-BREAK"');
+                self::assertNotContains($line, $title3s, 'CHANGELOG line ' . $index . ' MUST have unique title3');
+                $title3s[] = $line;
+                $title2Empty = false;
+                $inTitle3 = true;
+                $title3Empty = true;
+                $wasEmptyLine = false;
+
+                $needEmptyLine = true;
+
+                continue;
+            }
+
+            $wasEmptyLine = false;
+            if ($inTitle3) {
+                self::assertMatchesRegularExpression('/^[ -] /', $line, 'CHANGELOG entry at line ' . $index . ' MUST start with "- " or "  " for multiline content');
+                $title3Empty = false;
+
+                continue;
+            }
+
+            if ('- Empty release' === $line) {
+                $needEmptyLine = true;
+                $inTitle3 = false;
+                $title2Empty = false;
+
+                continue;
+            }
+
+            if ('First version' === $line) {
+                $finalWords = true;
+                $needEmptyLine = true;
+
+                continue;
+            }
+
+            self::assertFalse('CHANGELOG entry at line ' . $index . ' is not expected');
+        }
+
+        self::assertTrue($finalWords, 'CHANGELOG MUST contains "First version"');
+    }
+
+    /**
+     * @dataProvider provideChangedServicesWithoutChangelog
+     */
+    public function testChangelogEntryForService(string $service, string $base, $isCommentOnly)
+    {
+        if (!$isCommentOnly) {
+            self::fail('Missing CHANGELOG entry for package ' . $service);
+        }
+
+        $changelog = explode("\n", file_get_contents($base . '/CHANGELOG.md'));
+        $nrSection = false;
+
+        foreach ($changelog as $line) {
+            if ('## NOT RELEASED' === $line) {
+                $nrSection = true;
+
+                continue;
+            }
+            if (!$nrSection) {
+                continue;
+            }
+            if (0 === strpos($line, '## ')) {
+                break;
+            }
+
+            if ('- AWS enhancement: Documentation updates.' === $line) {
+                $this->expectNotToPerformAssertions();
+
+                return;
+            }
+        }
+
+        self::fail('Missing CHANGELOG entry for package ' . $service);
+    }
+
+    public function provideChangedlogFiles(): iterable
```

Comment:
Data providers should be static methods. Non-static ones are deprecated in latest PHPUnit.

Reply:
thanks